src/parser.rs

```diff
@@ -615,8 +615,7 @@ impl<'a> Scanner<'a> {
         #[cold]
         fn badchar(slf: &mut Scanner, ix: usize) -> Span {
             let ch = slf.buffer[ix];
-            let diag = Diagnostic::BadCharacter(ix, ch);
-            slf.diagnostics.push((slf.statement_index, diag));
+            slf.diag(Diagnostic::BadCharacter(ix, ch));
```

Comment:
pre-`badchar`, I was doing this directly because the `self.diag` call could change `self.buffer`, preventing the compiler from keeping the buffer pointer in a register in `get_raw`.  Now that we've got the tail recursion hack, there's no method calls in the actual loop, so this may no longer be needed